### PR TITLE
Repeatable tests

### PR DESCRIPTION
**Pull request by cubisttriangle**
Create ReplayClient, which records the output of a real request and saves it to a file. Subsequent, identical requests will return the stored response, rather than making a real network request.

Here are the steps for creating a mock api test:
- Create a test which makes a finnhub api call using a real key. This will save a file in the 'tests' folder with a name matching the query minus the root and having the extension '.replay'. That way we can have a test for any endpoint. When the file is saved, the real key in the filename and file contents will be replaced with a dummy key.
- Now that there is a cached response file, we can switch to using a dummy key. The cached response will be loaded.
- Copy the '{:#?}' output of the deserialized json and save to a file whose base filename matches the replay file and whose extensionis '.expected'.
- Now the deserialized output from a cached response can be compared the expected output to the cached response.

#8

### CODE REVIEW

**Comment by cubisttriangle** [comment]
Currently the only failing test is the doc-test which uses a real api key, so this must mean there is an issue with the environment setup. It definitely took me a few tries when I was getting everything setup on my fork.

**Comment by cubisttriangle** [comment]
Alright, think I'm done with this for now. Let's discuss when you get a chance. No rush of course.

File: src/lib.rs [hunk 1]
```diff
@@ -27,142 +27,149 @@
...
 mod test {
     // `cargo t -- --nocapture` to test
     use super::client::*;
     use super::utils::*;
-    use crate::types::{MarketNewsCategory, ProfileToParam};
+    use super::types::*;
+
+
+    /// Macro to generate client and run api test. Helps remove a bunch of boilerplate.
+    macro_rules! api_test {
```
Comment (henryboisdequin) [inline src/lib.rs:48]: Nice macro!

File: src/utils.rs [hunk 2]
```diff
@@ -10,3 +11,37 @@ pub fn get_test_api_key() -> String {
...
+    let mut dst = std::fs::File::create(filename).unwrap();
+    dst.write(new_data.as_bytes()).unwrap();
+}
+
+/// Replace the file extension of a replay file with that of an expected. Not very robust.
+#[cfg(test)]
+pub fn load_expected_from_replay_filename(replay_filename: String) -> String {
+    let expected_filename = replay_filename.replace(".replay", ".expected");
+    std::fs::read_to_string(expected_filename).unwrap().trim().into()
+}
```
Comment (henryboisdequin) [inline src/utils.rs]: A new line would be preferred!

**Review by henryboisdequin — APPROVED** [review]
Thanks for this PR, looks amazing. I didn't know you could test in such a way. This is a great milestone for #8, just need to solve the todo you added:

> // TODO: Figure out a real test. Also, this test always seems to terminate before completion, so figure that out.

File: src/utils.rs (hunk 2 above)
Comment (cubisttriangle) [reply src/utils.rs]: Done. Need a git commit hook for that or something.

Comment (henryboisdequin) [reply src/utils.rs]: Yeah, that would be great

**Pull request merged**